### DISCUSSION TITLE
Java EE and JSF

### DISCUSSION THREAD

**Discussion by manimovassagh** [cypress-io/cypress#27945] [Questions and Help]
Hello everybody,
I am in a project that we use Java EE and JSF primefaces. because it has no rest API and is a fullstack java app , automatic cypress waiting most of the times failes and I have to explicitly give cy.wait() in End to End test for many cases. is there any solution that i be able to observe the DOM for stability and finished any update and then run the next line test?
for example I have a upload csv button and i give the file. how can i opberve the dom that when this is finished and dom is getting stable?
Thanks

**Comment by portyu9** [comment]
Hi @manimovassagh 

There isn't a generic Cypress command for "wait until the DOM is stable", and I would avoid implementing one with a `MutationObserver`.

A DOM can appear stable for 100 ms and then be changed by another async response, timer, PrimeFaces AJAX update, etc. A "quiet DOM" therefore doesn't necessarily mean the application is finished.

Instead, wait for the **observable condition that means the operation is complete**.

For example, if the PrimeFaces/JSF action makes an AJAX request, intercept that request:

```js
cy.intercept('POST', '**/your-page*').as('update')

cy.get('[data-cy="upload"]').selectFile('file.csv')

cy.wait('@update')
cy.get('[data-cy="loading"]').should('not.exist')
cy.get('[data-cy="result"]').should('be.visible')
```

Adjust the route/selector to match your application.

If there isn't a convenient request to intercept, assert directly on the state the user cares about:

```js
cy.get('.ui-blockui').should('not.exist')
cy.contains('Upload completed').should('be.visible')
```

Cypress retries queries and assertions automatically, so they naturally handle variable server/rendering time without a fixed sleep.

For a full-page JSF form submission, similarly wait for the resulting URL or page content rather than an arbitrary delay.

The key is to define "finished" in application terms—request completed, loading indicator disappeared, expected row appeared, button became enabled, etc.—rather than trying to detect generic DOM stability.

Please mark this answer as accepted if it helped, thank you.
